### DISCUSSION TITLE
GraphicsMagick support removed from MiniMagick

### DISCUSSION THREAD

**Discussion by bobfriesenhahn** [minimagick/minimagick#578] [General]
Someone brought my attention to a recent MiniMagick commit which removed GraphicsMagick support.  A key concern expressed is lack of a 'time' limit.  It was almost mentioned that the command variations (presumably between GraphicsMagick, the many variations of ImageMagick 6, and the many variations of ImageMagick 7) is causing hardship.

FYI, the GraphicsMagick oss-fuzz testing done by Google every day is done without any time limit feature yet the test needs to commit in 2 minutes, or oss-fuzz will report a bug.  Reasonably large input file sizes are accepted.  Limiting insane requests is entirely constrained via the other available resource limits, and checks in the code. Regardless, few timeout issues are reported.

Absent other limits, it is possible to create a GraphicsMagick request which could take days to complete.

What is necessary in order for MiniMagick to restore support for GraphicsMagick?  If I implement a time limit, would that be sufficient?

**Comment by janko** [comment]
MiniMagick did implement a time limit in Ruby that was used for both ImageMagick and GraphicsMagick commands, but I didn't want to maintain it anymore, I wanted to use `Open3.capture3` instead of reimplementing it with `Open3.popen3`. I also didn't want to maintain code for detection and changing between ImageMagick and GraphicsMagick, especially with IM6 vs IM7 differences that you mentioned.

Note that you can still use GraphicsMagick, it's just that not all MiniMagick features will be supported, as mentioned [here](https://github.com/minimagick/minimagick?tab=readme-ov-file#graphicsmagick). Is that not sufficient for your use cases?

Moving this into a discussion.

**Reply by bobfriesenhahn** [reply]
FYI, I am the GraphicsMagick maintainer, so I do not have any particular use cases.

It would be nice to remove impediments which keep GraphicsMagick from being properly supported.  Thus far I see that MiniMagick wants to rely on the the ImageMagick command-timeout feature and I see mention of its JSON output for image description.

**Reply by janko** [reply]
Sorry, I forgot to follow up on this. The GraphicsMagick-specific code that MiniMagick needed to maintain consisted of:

* the time limit (which you indicated might be wrong for MiniMagick to rely on)
* parsing `identify -verbose` output due to lack of JSON format
* differences in the `identify -format '%[EXIF:*]'` output (see [code](https://github.com/minimagick/minimagick/commit/718280cf0095e745fe8845faf0dc29d5ff378f44#diff-c46665b8d3decd2bbad74a0256c07f4a23d42436f82e4b3bf667d83b9e9b8f28L106-L109))

Could you provide me with an image that breaks IM's time limit, just so that I can verify this on my machine? If I can reproduce, then I'll probably need to revert to my own time limiting.

In that case, if GM adds support for JSON image data, then I can probably add the GM support back.

**Comment by bobfriesenhahn** [comment]
FYI, the ImageMagick time limit is based on polling. Its time limit implementation only works at all if the time is spent in one of its polling loops.  I just verified that if the time is spent in 3rd party code (in this case libjbig) that a request for a 1 second time limit waited for that 3rd party code to complete, and that took 50 seconds.  It is not difficult to find files which may take minutes, or a week to complete.

I recommend that MiniMagick retain its own time-limit code.

**Comment by bobfriesenhahn** [comment]
The attached file is a 27 byte JBIG file which takes a long time (e.g. 50 seconds) before it reports an error.  a ".jpg" extension has been added in order to satisfy Github's upload requirements.  This is only useful if ImageMagick includes JBIG support.

![clusterfuzz-testcase-minimized-coder_JBIG_fuzzer-6622743017422848 jbig](https://github.com/user-attachments/assets/9a90d8b6-921e-4f77-9ba9-06fbd0dea215)
